File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    ans=0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def height(node):
            if node==None: return 0
            lh=height(node.left)
            rh=height(node.right)
            dia=(lh+rh)
            self.ans=max(self.ans,dia)
            return max(lh,rh)+1
        height(root)
        return self.ans

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return ans;
    }
    int height(TreeNode* node){
        if (node==NULL){
            return  0;
        }
        int lh=height(node->left);
        int rh=height(node->right);
        int dia=lh+rh;
        ans=max(ans,dia);
        return max(lh,rh)+1;
    }
};